private-templates-service/server/src/db/providers/InMemoryDBProvider.ts

```diff
@@ -0,0 +1,249 @@
+import {Queries, Collections, StorageProvider, Interface} from './StorageProvider'
```

Comment:
While this is a very full implementation of the provider, it could probably be as simple as this:

`module Interface {

    export interface IUser  {
        team: string[],
        org: string[],
        email: string
    }

    export interface ITemplateInstance {
        json: string,
        version: string
    }

    export interface ITemplate {
        instances: ITemplateInstance[],
        tags: string[],
        owner: string,
        createdAt: Date,
        updatedAt: Date,
        isPublished: boolean
    }
}


export class InMemoryDBProvider implements StorageProvider {

    users: Array<Interface.IUser> = [];
    templates: Array<Interface.ITemplate> = [];


    async  insertTemplate(template: Collections.Template) : Promise<void> {
        if(this.templates.includes(template)) {
            throw new Error("template already exists")
        }
        this.templates.push(template)
    }

    async insertUser(user: Collections.User) : Promise<void> {
        if(this.users.includes(user)) {
            throw new Error("user already exists")
        }
        this.users.push(user)
    }

    getTemplate(query: Queries.TemplateQuery) : Promise<Collections.Template[]> {
        let tmpTemplates = this.templates

        // logic to filter query

        return Promise.resolve(tmpTemplates)
    }


    getUser(query: Queries.UserQuery) : Promise<Collections.User[]> {
        let tmpUsers = this.users

        // logic to filter query

        return Promise.resolve(tmpUsers)
    }
}`

Reply:
Actually what it looks like you have implemented is a a file storage database rather than in memory